KAFKA-18601: Assume a baseline of 3.3 for server protocol versions

ijuma opened the pull request:
3.3.0 was the first KRaft release that was deemed production-ready and also when KIP-778 (KRaft to KRaft upgrades) landed. Given that, it's reasonable for 4.x to only support upgrades from 3.3.0 or newer (the metadata version also needs to be set to "3.3" or newer before upgrading).

Noteworthy changes:
1. `AlterPartition` no longer includes topic names, which makes it possible to simplify `AlterParitionManager` logic.
2. Metadata versions older than `IBP_3_3_IV3` have been removed and `IBP_3_3_IV3` is now the minimum version.
3. `MINIMUM_BOOTSTRAP_VERSION` has been removed.
4. Removed `isLeaderRecoverySupported`, `isNoOpsRecordSupported`, `isKRaftSupported`, `isBrokerRegistrationChangeRecordSupported` and `isInControlledShutdownStateSupported` - these are always `true` now. Also removed related conditional code.
5. Removed default metadata version or metadata version fallbacks in multiple places - we now fail-fast instead of potentially using an incorrect metadata version.
6. Update `MetadataBatchLoader.resetToImage` to set `hasSeenRecord` based on whether image is empty - this was a previously existing issue that became more apparent after the changes in this PR.
7. Remove `ibp` parameter from `BootstrapDirectory`
8. A number of tests were not useful anymore and have been removed.

I will update the upgrade notes via a separate PR as there are a few things that need changing and it would be easier to do so that way.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma committed:
KAFKA-18601: Assume a baseline of 3.3 for server protocol versions

ijuma commented on tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java:
```diff
@@ -128,18 +126,6 @@ public void testReassignment() throws Exception {
-        // Test reassignment when the IBP is on an older version which does not use
-        // the `AlterPartition` API. In this case, the controller will register individual
-        // watches for each reassigning partition so that the reassignment can be
-        // completed as soon as the ISR is expanded.
```
Has the code for this fallback been removed? If not, do we have a JIRA ticket for it? If not, I'll file one.

ijuma replied:
@cmccabe Do you know?

chia7712 replied:
It seems to me the behavior is related to zk only, and in kraft mode `AlterPartition` should be supported totally. Hence, it is fine to remove the test.

BTW, the test case for kraft is unnecessary - I overlook the scenario in merging #15675

ijuma replied:
Makes sense. To confirm, you're saying we can remove this test _and_ we don't need any other follow-up since the comment was for the zk controller. Right?

chia7712 replied:
>  you're saying we can remove this test and we don't need any other follow-up since the comment was for the zk controller. Right?

we are on the same page!

ijuma commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -1813,7 +1813,7 @@ class Partition(val topicPartition: TopicPartition,
     debug(s"Submitting ISR state change $proposedIsrState")
     val future = alterIsrManager.submit(
-      new TopicIdPartition(topicId.getOrElse(Uuid.ZERO_UUID), topicPartition),
+      new org.apache.kafka.server.common.TopicIdPartition(topicId.getOrElse(throw new IllegalStateException("Topic id not set")), topicPartition.partition),
```
It's not safe to use `Uuid.ZERO_UUID` with the changes in this PR and hence we now throw an exception is the topic is is not known. @jolshan Is this OK?

jolshan replied:
I think this makes sense. I'm not sure exactly though what the error handling is here when an error is thrown. It seems like the futures here are not always checked. (Also wondering if we want to include more information about the partition in the error message)

dajac replied:
I wonder whether we should make the topic id mandatory (not an Option) in the Partition now. It is always set in KRaft anyway.

ijuma replied:
@dajac I agree with you, but I thought that should be done in a separate PR as I think it will require some refactoring given the way various things interact right now. And perhaps it can be done in trunk only (no backport to 4.0 required).

ijuma replied:
> I think this makes sense. I'm not sure exactly though what the error handling is here when an error is thrown. It seems like the futures here are not always checked. (Also wondering if we want to include more information about the partition in the error message)

My assumption is that the topic id is actually always set when this method is called, so the error thrown is just a way to make sure we are not missing something. Once we refactor things, this will be more obvious, but it would be nice to avoid blocking 4.0 on that.

ijuma replied:
> if we want to include more information about the partition in the error message

Good idea - I will add the `topicPartition` to the error message.

jolshan replied:
> so the error thrown is just a way to make sure we are not missing something. 

Yeah, that makes sense. The main thing I was wanting to check is that the error doesn't get swallowed somewhere. 👍

ijuma commented on metadata/src/main/java/org/apache/kafka/image/node/MetadataNode.java:
```diff
@@ -53,7 +53,8 @@ default void print(MetadataNodePrinter printer) {
             MetadataNode child = child(name);
-            child.print(printer);
+            if (child != null)
+                child.print(printer);
```
Fixes a NPE I noticed while debugging a failing test.

chia7712 replied:
In reading the code of child, I notice that it seems the `AclsImageByIdNode` is misplaced. 

```
    private static final Map<String, Function<MetadataImage, MetadataNode>> CHILDREN = Map.of(
         ...
         // AclsImageByIdNode should be replaced by AclsImageNode
        AclsImageNode.NAME, image -> new AclsImageByIdNode(image.acls()),
        ScramImageNode.NAME, image -> new ScramImageNode(image.scram()),
        DelegationTokenImageNode.NAME, image -> new DelegationTokenImageNode(image.delegationTokens())
    );
```
https://github.com/apache/kafka/blob/48283ad2e54133ebb604769dc8a72706e07f36e7/metadata/src/main/java/org/apache/kafka/image/node/MetadataImageNode.java#L46

and `AclsImage#toString` has similar issue.
```
    @Override
    public String toString() {
        // AclsImageByIdNode should be replaced by AclsImageNode
        return new AclsImageByIdNode(this).stringify();
    }
```

https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/image/AclsImage.java#L78

@ijuma @mumrah WDYT?

ijuma replied:
Interesting. Shall we file a separate JIRA ticket for that? Perhaps we can wait for @mumrah to confirm.

mumrah replied:
Hm, it's definitely a bug, but I don't think it relates to this PR. The *Node classes are used by the metadata shell and some toString methods, so they are just informational. I think the exact side effect is that acls would appear at the wrong level of the metadata shell "tree". 

I suggest we file a separate issue and fix it for 4.0.

chia7712 replied:
@ijuma @mumrah thanks for your response. open https://issues.apache.org/jira/browse/KAFKA-18803 for it

ijuma commented on metadata/src/main/java/org/apache/kafka/metadata/bootstrap/BootstrapDirectory.java:
```diff
@@ -86,12 +84,7 @@ BootstrapMetadata readFromConfiguration() {
             return BootstrapMetadata.fromVersion(MetadataVersion.latestProduction(), "the default bootstrap");
         }
         MetadataVersion version = MetadataVersion.fromVersionString(ibp.get());
-        if (version.isLessThan(MINIMUM_BOOTSTRAP_VERSION)) {
```
@cmccabe Is it important to automatically fallback if older metadata versions are received here or can we throw an error? It's harder to fallback (but not impossible) since I removed older metadata versions.

chia7712 replied:
the `ibp` is always empty after the config is removed by #18566

I had opened https://issues.apache.org/jira/browse/KAFKA-18740 to cleanup it. Maybe this PR can include the cleanup as well?

ijuma replied:
Sure, will push a commit with this change.

ijuma replied:
I guess this means the answer to my question is that the fallback is not important.

cmccabe replied:
yes, we're getting rid of the fallback

ijuma commented on metadata/src/test/java/org/apache/kafka/image/ImageDowngradeTest.java:
```diff
@@ -77,59 +77,6 @@ static ApiMessageAndVersion metadataVersionRecord(MetadataVersion metadataVersio
-     * Test downgrading to a MetadataVersion that doesn't support FeatureLevelRecord.
-     */
-    @Test
-    public void testPremodernVersion() {
```
We don't support these anymore, so remove it.

ijuma commented on metadata/src/test/java/org/apache/kafka/image/ImageDowngradeTest.java:
```diff
@@ -77,59 +77,6 @@ static ApiMessageAndVersion metadataVersionRecord(MetadataVersion metadataVersio
-     * Test downgrading to a MetadataVersion that doesn't support inControlledShutdown.
-     */
-    @Test
-    public void testPreControlledShutdownStateVersion() {
```
We now always have the controlled shutdown state.

ijuma committed:
Fix a few tests and tweak error messages

junrao commented on core/src/main/scala/kafka/server/AlterPartitionManager.scala:
```diff
@@ -224,53 +215,37 @@ class DefaultAlterPartitionManager(
-      val topicData = new AlterPartitionRequestData.TopicData()
-        .setTopicName(topicName)
-        .setTopicId(topicId)
+      val topicData = new AlterPartitionRequestData.TopicData().setTopicId(topicId)
```
Could we adjust the comment on topic name above?

junrao commented on server-common/src/main/java/org/apache/kafka/server/common/UnitTestFeatureVersion.java:
```diff
@@ -321,7 +321,7 @@ public Map<String, Short> dependencies() {
      */
     public enum FV7 implements FeatureVersion {
-        UT_FV7_0(0, MetadataVersion.MINIMUM_KRAFT_VERSION, Collections.singletonMap(MetadataVersion.FEATURE_NAME, MetadataVersion.IBP_3_7_IV0.featureLevel())),
+        UT_FV7_0(0, MetadataVersion.MINIMUM_VERSION, Collections.singletonMap(MetadataVersion.FEATURE_NAME, MetadataVersion.IBP_3_7_IV0.featureLevel())),
```
The error message in FeatureTest with respect to 3.0 needs to be adjusted. 
```
            "Feature UNIT_TEST_VERSION_7 has default FeatureVersion UT_FV7_0 when MV=3.0-IV1 with " +
                "MV dependency 3.7-IV0 that is behind its bootstrap MV 3.0-IV1.");
```

ijuma replied:
Interesting - that test is passing, so I'll have to check why.

junrao replied:
The test passes since the error message is only used if an expected exception is not thrown.

ijuma replied:
I'll change the test to check the error message so that this doesn't happen in the future.

junrao commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -220,18 +187,6 @@ public boolean isElrSupported() {
-        return this.featureLevel > 0;
-    }
-
-    public boolean isBrokerRegistrationChangeRecordSupported() {
```
The comment in BrokerRegistrationTrackerTest.java with respect to 3.0 needs to be adjusted.
`        // No calls are made because MetadataVersion is 3.0-IV1 initially`

junrao commented on metadata/src/test/java/org/apache/kafka/metadata/bootstrap/BootstrapDirectoryTest.java:
```diff
@@ -81,18 +81,18 @@ public void testReadFromEmptyConfiguration() throws Exception {
-            assertEquals(BootstrapMetadata.fromVersion(MetadataVersion.MINIMUM_BOOTSTRAP_VERSION,
-                    "the minimum version bootstrap with metadata.version 3.3-IV0"),
+            assertEquals(BootstrapMetadata.fromVersion(MetadataVersion.MINIMUM_VERSION,
+                    "the minimum version bootstrap with metadata.version 3.3-IV3"),
```
The usage of 3.0 in the next line and in `testReadFromConfigurationFile` need to be adjusted.

ijuma replied:
Subsequent commits simplified this file and I think the issue you outlined is no longer present.

junrao commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -44,25 +44,6 @@
-    IBP_3_3_IV1(5, "3.3", "IV1", true),
-
-    // In KRaft mode, use BrokerRegistrationChangeRecord instead of UnfenceBrokerRecord and FenceBrokerRecord.
-    IBP_3_3_IV2(6, "3.3", "IV2", true),
```
The references of the old MV in AclsImage  need to be adjusted.

ijuma committed:
Minor improvements, address review comments and fix tests

chia7712 commented on metadata/src/test/java/org/apache/kafka/metadata/bootstrap/BootstrapMetadataTest.java:
```diff
@@ -126,7 +125,7 @@ public void testFeatureLevelForFeature() {
             new ApiMessageAndVersion(new FeatureLevelRecord().
                 setName(FEATURE_NAME).
-                setFeatureLevel(IBP_3_0_IV1.featureLevel()), (short) 0));
+                setFeatureLevel((short) 1), (short) 0));
```
This feature version is not supported any more, so we should either remove or modify the test.

ijuma replied:
This is intended - see the name of the field (RECORDS_WITH_OLD_METADATA_VERSION). I did update this to refer to a constant.

chia7712 commented on metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java:
```diff
@@ -328,15 +314,10 @@ private ApiError updateMetadataVersion(
             newVersion = MetadataVersion.fromFeatureLevel(newVersionLevel);
         } catch (IllegalArgumentException e) {
-            return invalidMetadataVersion(newVersionLevel, "Unknown metadata.version.");
+            return invalidMetadataVersion(newVersionLevel, "Valid versions are from "
```
Should we enhance the error message for `fromFeatureLevel` as well?

ijuma replied:
I think I did this in a subsequent commit, perhaps after you did your review.

ijuma committed:
Formatting fixes

ijuma commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -170,41 +140,17 @@ public void testPrevious() {
-    }
-
-    @Test
-    public void testKRaftVersions() {
```
This did not seem to add much value after the changes in this PR.

chia7712 commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1379,16 +1367,10 @@ void handleBrokerFenced(int brokerId, List<ApiMessageAndVersion> records) {
-                    setId(brokerId).setEpoch(brokerRegistration.epoch()),
-                    (short) 0));
-        }
+        records.add(new ApiMessageAndVersion(new BrokerRegistrationChangeRecord().
```
Could you please update comment of `handleBrokerFenced` according to this change?

chia7712 commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1424,15 +1406,10 @@ void handleBrokerUnregistered(int brokerId, long brokerEpoch,
-            records.add(new ApiMessageAndVersion(new UnfenceBrokerRecord().setId(brokerId).
-                setEpoch(brokerEpoch), (short) 0));
-        }
+        records.add(new ApiMessageAndVersion(new BrokerRegistrationChangeRecord().
```
ditto

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ActivationRecordsGenerator.java:
```diff
@@ -137,10 +137,10 @@ static ControllerResult<Void> recordsForNonEmptyLog(
         }
 
-        if (curMetadataVersion.equals(MetadataVersion.MINIMUM_KRAFT_VERSION)) {
+        if (curMetadataVersion.equals(MetadataVersion.MINIMUM_VERSION)) {
```
MetadataVersion.MINIMUM_VERSION now supports metadata.version feature level record. So, the following log message could be confusing.

ijuma replied:
Good catch. We have two options: (1) Remove this log or (2) Pass a parameter indicating whether `metadata.version` feature level record is present. Do you have an opinion on which way is better? I'll also take a closer look to understand whether the second option is worth it.

ijuma replied:
I went ahead and removed the log. @cmccabe Let me know if you think we should do something different.

chia7712 commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -1652,9 +1652,7 @@ private void registerWriteNoOpRecord(long maxIdleIntervalNs) {
-                if (featureControl.metadataVersion().isNoOpRecordSupported()) {
-                    records.add(new ApiMessageAndVersion(new NoOpRecord(), (short) 0));
-                }
+                records.add(new ApiMessageAndVersion(new NoOpRecord(), (short) 0));
```
Maybe we can streamline the code?
```
() -> ControllerResult.of(List.of(new ApiMessageAndVersion(new NoOpRecord(), (short) 0)), false),
```

ijuma replied:
The new version would make `records` immutable while the current code exposes the mutable `records` via `ControllerResult.records()`. Do you know if this change is safe? My cursory look indicates it may be ok, but I would have to look at this more carefully. Given that, I think we should also do this via a different PR (to avoid including a potential behavior change in the middle of this clean-up PR).

chia7712 replied:
> Do you know if this change is safe? My cursory look indicates it may be ok, but I would have to look at this more carefully. 

Since other similar cases pass immutable lists to `ControllerResult`, it seems safe to do so here as well.

> I think we should also do this via a different PR (to avoid including a potential behavior change in the middle of this clean-up PR).

make sense to me

cmccabe replied:
@ijuma , we do not mutate the records list that gets returned from these functions later on. @chia7712 is correct that we return immutable objects in many other cases. Unfortunately, due to Java's lack of `const` I guess that isn't obvious.

ijuma replied:
OK, I'll include this change in my next push.

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:
```diff
@@ -293,11 +293,7 @@ public ApiMessageAndVersion toRecord(ImageWriterOptions options) {
-            } else {
-                options.handleLoss("the inControlledShutdown state of one or more brokers");
-            }
+            registrationRecord.setInControlledShutdown(true);
```
Maybe we can set the flag in line#294. 
```
        RegisterBrokerRecord registrationRecord = new RegisterBrokerRecord().
            setBrokerId(id).
            setRack(rack.orElse(null)).
            setBrokerEpoch(epoch).
            setIncarnationId(incarnationId).
            setFenced(fenced).
            setInControlledShutdown(inControlledShutdown);
```

chia7712 reviewed (commented):
@ijuma thanks for this patch. overall LGTM. some minor comments remains.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -495,7 +495,9 @@ BrokerFeature processRegistrationFeature(
-        int defaultVersion = feature.name().equals(MetadataVersion.FEATURE_NAME) ? 1 : 0; // The default value for MetadataVersion is 1 not 0.
+        // The default value for MetadataVersion changes over time while other features start at `0`
+        int defaultVersion = feature.name().equals(MetadataVersion.FEATURE_NAME) ?
+            MetadataVersion.MINIMUM_VERSION.featureLevel() : 0;
```
@cmccabe : Should the finalized MetadataVersion always be present in any release upgradable to 4.0?

cmccabe replied:
If we're dropping support for everything before 3.3 we can always count on having a FeatureRecord for `metadata.version`. However doing this may require a lot of code changes in tests, since many junit tests may initialize with an empty log. (Or maybe not... I would have to check?)

ijuma replied:
I can try. We have one test failing only in CI due to the minimum being used, perhaps this will help.

ijuma replied:
I made this change.

ijuma committed:
Minor tweak

junrao commented on metadata/src/main/java/org/apache/kafka/image/FeaturesImage.java:
```diff
@@ -56,7 +53,7 @@ public FeaturesImage(
     public boolean isEmpty() {
         return finalizedVersions.isEmpty() &&
-            metadataVersion.equals(MetadataVersion.MINIMUM_KRAFT_VERSION);
+            metadataVersion.equals(MetadataVersion.MINIMUM_VERSION);
```
Hmm, metadata version 3.3 can be set explicitly in KRaft. Is this still correct @cmccabe ?

ijuma replied:
This is a good question, I assumed both conditions together would still mean empty, but this assumption may be wrong.

junrao replied:
`isEmpty()` is only used in `MetadataLoaderTest`. So, we can address this in a followup jira. But I am not sure if `isEmpty()` will ever be false given that metadata version is required in the metadata log and probably the snapshot.

ijuma replied:
It turns out that this had to be fixed for all the tests to pass. Addressed in the latest commit.

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kafka-18601-assume-baseline-3.3-for-server

* apache-github/trunk:
  KAFKA-18366 Remove KafkaConfig.interBrokerProtocolVersion (#18820)
  KAFKA-18658 add import control for examples module (#18812)
  MINOR: Java version and TLS documentation improvements (#18822)
  KAFKA-18743 Remove leader.imbalance.per.broker.percentage as it is not supported by Kraft (#18821)
  KAFKA-18225 ClientQuotaCallback#updateClusterMetadata is unsupported by kraft (#18196)
  KAFKA-18441: Remove flaky tag on KafkaAdminClientTest#testAdminClientApisAuthenticationFailure (#18847)
  MINOR: fix KStream#to incorrect javadoc (#18838)
  KAFKA-18763: changed the assertion statement for acknowledgements to include only successful acks (#18846)
  MINOR: Accept specifying consumer group assignors by their short names (#18832)

junrao commented on server-common/src/main/java/org/apache/kafka/server/common/Feature.java:
```diff
@@ -155,18 +155,16 @@ public FeatureVersion fromFeatureLevel(short level,
         Short metadataVersion = features.get(MetadataVersion.FEATURE_NAME);
 
-        if (feature.featureLevel() >= 1 && (metadataVersion == null || metadataVersion < MetadataVersion.IBP_3_3_IV0.featureLevel()))
+        if (feature.featureLevel() >= 1 && (metadataVersion == null || metadataVersion < MetadataVersion.MINIMUM_VERSION.featureLevel()))
```
It seems that we could remove this statement completely since the callers will always have metadataVersion set to >= 3.3. Is that correct @jolshan ?

ijuma replied:
Since we are not actually using `MetadataVersion` here, it seems possible for callers to pass invalid values. That is why I kept the validation.

junrao commented on server-common/src/main/java/org/apache/kafka/server/common/Feature.java:
```diff
@@ -298,7 +296,7 @@ public static void validateDefaultValueAndLatestProductionValue(
         for (MetadataVersion metadataVersion: MetadataVersion.values()) {
             // Only checking the kraft metadata versions.
-            if (metadataVersion.compareTo(MetadataVersion.MINIMUM_KRAFT_VERSION) < 0) {
+            if (metadataVersion.compareTo(MetadataVersion.MINIMUM_VERSION) < 0) {
```
This is no longer needed since all passed in metadataVersion are >= `MetadataVersion.MINIMUM_VERSION`.

junrao commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -552,24 +535,10 @@ class AlterPartitionManagerTest {
+  def provideLeaderRecoveryState(): JStream[Arguments] = {
     JStream.of(
-      // Supports KIP-903: include broker epoch in AlterPartition request
-      IBP_3_5_IV1,
```
Should we change to test both IBP_3_3_IV3 and IBP_3_5_IV1 here, @CalvinConfluent ?

ijuma replied:
I removed that method altogether since none of the tests used it anymore.

ijuma committed:
Remove `ibp` parameter from `BootstrapDirectory`

junrao commented on core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:
```diff
@@ -380,7 +380,7 @@ public void testUpdateBrokerConfigNotAffectedByInvalidConfig() {
          // Must be at greater than 1MB per cleaner thread, set to 2M+2 so that we can set 2 cleaner threads.
          serverProperties = {@ClusterConfigProperty(key = "log.cleaner.dedupe.buffer.size", value = "2097154")},
-         metadataVersion = MetadataVersion.IBP_3_3_IV0
+         metadataVersion = MetadataVersion.IBP_3_9_IV0
```
Hmm, why do we choose IBP_3_9_IV0 here?

ijuma replied:
Since we cannot use `MINIMUM_VERSION` (see my other comment), I picked a recent version so there is less chance of this code changing. If it's confusing, I can change it to `IBP_3_3_IV3`.

ijuma committed:
Address a couple of minor review comments

junrao commented on core/src/test/scala/integration/kafka/coordinator/transaction/ProducerIntegrationTest.scala:
```diff
@@ -55,7 +55,7 @@ import scala.jdk.CollectionConverters._
 
   @ClusterTests(Array(
-    new ClusterTest(metadataVersion = MetadataVersion.IBP_3_3_IV0)
+    new ClusterTest(metadataVersion = MetadataVersion.IBP_3_3_IV3)
```
Should we use MINIMUM_VERSION to avoid keeping changing it in the future?

ijuma replied:
I had the same idea, but unfortunately it's not allowed:

> annotation argument needs to be a constant; found: org.apache.kafka.server.common.MetadataVersion.MINIMUM_VERSION

ijuma commented:
@chia7712 I addressed your comments except one (and left a reply explaining why).

junrao commented on metadata/src/test/java/org/apache/kafka/metadata/bootstrap/BootstrapDirectoryTest.java:
```diff
@@ -81,18 +81,17 @@ public void testReadFromEmptyConfiguration() throws Exception {
-                new BootstrapDirectory(testDirectory.path(), Optional.of("3.3-IV2")).read());
+            assertEquals(BootstrapMetadata.fromVersion(MetadataVersion.IBP_3_3_IV3,
+                    "the configured bootstrap with metadata.version 3.3-IV3"),
+                new BootstrapDirectory(testDirectory.path(), Optional.of("3.3-IV3")).read());
```
Should we change `3.0-IV0` in `testReadFromConfigurationFile` below?

ijuma replied:
I think the code you're referencing was removed in a subsequent commit.

junrao commented on metadata/src/test/java/org/apache/kafka/metadata/bootstrap/BootstrapMetadataTest.java:
```diff
@@ -126,13 +126,13 @@ public void testFeatureLevelForFeature() {
             () -> BootstrapMetadata.fromRecords(RECORDS_WITH_OLD_METADATA_VERSION, "quux"));
-        assertEquals("Bootstrap metadata.version before 3.3-IV0 are not supported. Can't load " +
-            "metadata from quux", exception.getMessage());
+        assertEquals("No MetadataVersion with feature level 1. Valid feature levels are from 7 to 25.",
```
Could we use defined constants for feature level to avoid keeping changing it in the future?

ijuma replied:
It's a good question and I wondered the same. One challenge is that the test ends up pretty similar to the actual code. But I don't feel strongly. Perhaps a middle ground is to use the constant for the latest version - which changes frequently. The minimum version changes very infrequently and it's probably good to find out if we accidentally change it. Thoughts?

ijuma replied:
I changed it to include constants for both cases.

ijuma commented on metadata/src/test/java/org/apache/kafka/metadata/bootstrap/BootstrapDirectoryTest.java:
```diff
@@ -74,39 +73,21 @@ public void testReadFromEmptyConfiguration() throws Exception {
-    }
-
-    @Test
-    public void testReadFromConfiguration() throws Exception {
```
`testReadFromConfiguration` is the same as `testReadFromEmptyConfiguration` once the `MetadataVersion` parameter is removed.

junrao reviewed (commented):
@ijuma : Thanks for the PR. Made through half of it. Left a few comments.

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kafka-18601-assume-baseline-3.3-for-server

* apache-github/trunk:
  KAFKA-18770 close the RM created by testDelayedShareFetchPurgatoryOperationExpiration (#18853)

ijuma committed:
Address review comments

ijuma commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -1524,31 +1507,6 @@ public void close() throws Exception {
             "test bootstrap");
 
-    @Test
-    public void testUpgradeFromPreProductionVersion() throws Exception {
```
Upgrades from pre production versions are no longer supported.

ijuma committed:
Replace IBP_3_3_IV3 with MINIMUM_VERSION where applicable

ijuma committed:
Set `topicId` a bit more consistently in `PartitionTest`

ijuma commented:
@junrao Thanks for the review. I addressed your comments and/or replied within the relevant thread. I also pushed a couple of commits for changes I had in progress: (1) Use `MINIMUM_VERSION` in more places (to reduce future changes) and (2) Set topic id more consistently in `PartitionTest`. Finally, I also merged from trunk hoping it fixes some of the test failures.

Let's see how the test results look after this.

m1a2st commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -368,7 +320,8 @@ public static MetadataVersion fromVersionString(String versionString) {
         }
         return Optional.ofNullable(IBP_VERSIONS.get(key)).orElseThrow(() ->
-            new IllegalArgumentException("Version " + versionString + " is not a valid version")
+            new IllegalArgumentException("Version " + versionString + " is not a valid version. The minimum version is " + MINIMUM_VERSION
+                + " and the maximum version is " + latestTesting())
```
This method is used in production code in `StorageTool.scala` and others, maybe we should not expose `latestTesting()` version, should  use `latestProduction()`.

ijuma replied:
If you check `MetadataVersion.testFromVersionString`, this code is accurate - `4.0-IV3` works even though it's a testing version still.

ijuma replied:
As you said, that seems to be inconsistent with the docs for `StorageTool` - perhaps we should file an issue to fix that in its own PR (assuming it's not a regression in this one, but I don't think the relevant code was changed).

m1a2st replied:
> As you said, that seems to be inconsistent with the docs for StorageTool - perhaps we should file an issue to fix that in its own PR (assuming it's not a regression in this one, but I don't think the relevant code was changed).

Make sense to me, I will file an issue to track it

m1a2st replied:
The `FeatureCommand.java` also have same issue, I addressed in same Jira https://issues.apache.org/jira/browse/KAFKA-18778

m1a2st commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -378,7 +331,8 @@ public static MetadataVersion fromFeatureLevel(short version) {
             }
         }
-        throw new IllegalArgumentException("No MetadataVersion with feature level " + version);
+        throw new IllegalArgumentException("No MetadataVersion with feature level " + version + ". Valid feature levels are from "
+            + MINIMUM_VERSION.featureLevel + " to " + latestTesting().featureLevel + ".");
```
ditto, we should use `latestProduction()`.

ijuma replied:
The code actually considers all metadata values - the error message s consistent with that.

m1a2st commented on tools/src/test/java/org/apache/kafka/tools/FeatureCommandTest.java:
```diff
@@ -250,66 +250,77 @@ public void testHandleDescribe() {
-    public void testHandleUpgrade() {
+    public void testHandleUpgradeToUnsupportedMetadataVersion() {
         Map<String, Object> namespace = new HashMap<>();
         namespace.put("metadata", "3.3-IV1");
```
nit: This `3.3-IV1` variable also can add to `MetadataVersionTestUtils` class.

ijuma replied:
I don't think we need to do that - the feature levels are opaque, but the metadata versions are clear and don't need a constant for usage in tests (they would need a constant within non test code but we don't use this in non test code).

m1a2st commented on server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:
```diff
@@ -44,25 +44,6 @@
  */
```
I think we should also update the java doc, replace old example like` IBP_0_10_0-IV0` to  `IBP_3_3_IV3`

ijuma replied:
Good point - I can also simplify the parsing code since we no longer have metadata versions starting with `0.`.

ijuma replied:
Done.

m1a2st reviewed (commented):
Thanks @ijuma for this PR, left some commments

ijuma committed:
Improve comments and remove code that is no longer needed

ijuma commented:
Besides the open discussion threads, there are two additional issues to consider/resolve:
1. Some tests in `PlaintextAdminIntegrationTest`are failing in the CI runs, but not locally. Unclear why that's the case.
2. `FeaturesDelta.replay` reads feature levels from the log and converts them to `MetadataVersion`. It's possible (but unlikely) that someone ended up with 3.3-IV0, 3.3-IV1, 3.3-IV2 in the log if using an unreleased version of Apache Kafka 3.3. We should consider automatically converting to `3.3-IV3` in such cases.

The failures have the following error:

> org.apache.kafka.server.fault.FaultHandlerException: quorumTestHarnessFaultHandler: Broker configuration does not support the cluster MetadataVersion: requirement failed: Multiple log directories (aka JBOD) are not supported in the current MetadataVersion 3.3-IV3. Need 3.7-IV2 or higher

Even though there is a log message just before saying:

> Formatting metadata directory /tmp/kafka-2205696658360646817 with metadata.version 4.0-IV3.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/FeatureControlManagerTest.java:
```diff
@@ -240,38 +241,29 @@ public void testReplayRecords() {
-                singletonMap(MetadataVersion.FEATURE_NAME, MetadataVersion.IBP_3_2_IV0.featureLevel()),
-                singletonMap(MetadataVersion.FEATURE_NAME, FeatureUpdate.UpgradeType.UNSAFE_DOWNGRADE),
-                true));
+            setFeatureLevel(MetadataVersion.IBP_3_4_IV0.featureLevel()));
```
Should we use MetadataVersion.MINIMUM_VERSION?

ijuma replied:
This was previously checking that the metadata version would change and using minimum would take it backwards. But I did notice an issue where we are actually using the same version versus a newer one - fixing that.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/FeatureControlManagerTest.java:
```diff
@@ -240,38 +241,29 @@ public void testReplayRecords() {
     @Test
     public void testCannotDowngradeToHigherVersion() {
-        FeatureControlManager manager = TEST_MANAGER_BUILDER1.build();
+        FeatureControlManager manager = new FeatureControlManager.Builder().
```
Could we use TEST_MANAGER_BUILDER1? This will make this test the same as `testCannotUnsafeDowngradeToHigherVersion` except for the unsafe flag, which is a bit easier to understand.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/FeatureControlManagerTest.java:
```diff
@@ -349,28 +344,27 @@ public void testCanUseUnsafeDowngradeIfMetadataChanged() {
             setQuorumFeatures(features(MetadataVersion.FEATURE_NAME,
-                MetadataVersion.IBP_3_0_IV1.featureLevel(), MetadataVersion.IBP_3_3_IV3.featureLevel())).
-            setMetadataVersion(MetadataVersion.IBP_3_3_IV0).
+                MetadataVersionTestUtils.IBP_3_0_IV1_FEATURE_LEVEL, MetadataVersion.MINIMUM_VERSION.featureLevel())).
```
We probably should set the min/max to `MetadataVersion.MINIMUM_VERSION` and `MetadataVersion.latestTesting()`.

ijuma replied:
The current code tests a slightly different path than the one you're suggesting. The error is thrown in a different place. I looked a bit more deeply though and perhaps your suggestion is closer to what will happen in production:

```java
public static Map<String, VersionRange> defaultSupportedFeatureMap(boolean enableUnstable) {
        Map<String, VersionRange> features = new HashMap<>(1);
        features.put(MetadataVersion.FEATURE_NAME, VersionRange.of(
                MetadataVersion.MINIMUM_VERSION.featureLevel(),
                enableUnstable ?
                    MetadataVersion.latestTesting().featureLevel() :
                    MetadataVersion.latestProduction().featureLevel()));
        for (Feature feature : Feature.PRODUCTION_FEATURES) {
            short maxVersion = enableUnstable ? feature.latestTesting() : feature.latestProduction();
            if (maxVersion > 0) {
                features.put(feature.featureName(), VersionRange.of(feature.minimumProduction(), maxVersion));
            }
        }
        return features;
    }
```

Changing it to your suggestion.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/FeatureControlManagerTest.java:
```diff
@@ -379,7 +373,7 @@ public void testCannotDowngradeBefore3_3_IV0() {
         Map<String, VersionRange> localSupportedFeatures = new HashMap<>();
         localSupportedFeatures.put(MetadataVersion.FEATURE_NAME, VersionRange.of(
-            MetadataVersion.IBP_3_0_IV1.featureLevel(), MetadataVersion.latestTesting().featureLevel()));
+            MetadataVersionTestUtils.IBP_3_0_IV1_FEATURE_LEVEL, MetadataVersion.latestTesting().featureLevel()));
```
We probably should set the min to MetadataVersion.MINIMUM_VERSION?

junrao commented on metadata/src/test/java/org/apache/kafka/image/FeaturesImageTest.java:
```diff
@@ -163,7 +163,7 @@ public void testEmpty() {
             FeaturesImage.EMPTY.metadataVersion()).isEmpty());
         assertFalse(new FeaturesImage(FeaturesImage.EMPTY.finalizedVersions(),
-            MetadataVersion.IBP_3_3_IV0).isEmpty());
+            MetadataVersion.IBP_3_4_IV0).isEmpty());
```
Since this can be any MV, could we just use MetadataVersion.MINIMUM_VERSION?

ijuma replied:
That causes the test to fail. This is related to your question of whether `FeatuesImage.isEmpty` should be changed not to include `metadataVersion.equals(MetadataVersion.MINIMUM_VERSION)`.

ijuma replied:
I updated `FeaturesImage.isEmpty` and also this test. Resolving.

junrao commented on metadata/src/test/java/org/apache/kafka/image/loader/MetadataLoaderTest.java:
```diff
@@ -310,7 +310,7 @@ public void testRemovePublisher() throws Exception {
                     new FeatureLevelRecord().
                         setName(MetadataVersion.FEATURE_NAME).
-                        setFeatureLevel(IBP_3_3_IV2.featureLevel()), (short) 0))));
+                        setFeatureLevel(IBP_3_3_IV3.featureLevel()), (short) 0))));
```
Since this can be any MV, could we just use MetadataVersion.MINIMUM_VERSION?

junrao commented on metadata/src/test/java/org/apache/kafka/image/loader/MetadataLoaderTest.java:
```diff
@@ -349,11 +349,11 @@ public void testLoadEmptySnapshot() throws Exception {
         }
         assertTrue(publishers.get(0).closed);
-        assertEquals(MetadataVersion.IBP_3_0_IV1,
+        assertEquals(IBP_3_3_IV3,
```
Could we just use MetadataVersion.MINIMUM_VERSION?

junrao commented on metadata/src/test/java/org/apache/kafka/image/loader/MetadataLoaderTest.java:
```diff
@@ -581,7 +581,7 @@ private void loadTestSnapshot(
                         singletonList(new ApiMessageAndVersion(new FeatureLevelRecord().
                                 setName(MetadataVersion.FEATURE_NAME).
-                                setFeatureLevel(IBP_3_3_IV1.featureLevel()), (short) 0)),
+                                setFeatureLevel(IBP_3_3_IV3.featureLevel()), (short) 0)),
```
Since this can be any MV, could we just use MetadataVersion.MINIMUM_VERSION?

junrao commented on metadata/src/test/java/org/apache/kafka/image/loader/MetadataLoaderTest.java:
```diff
@@ -511,7 +511,7 @@ public void testLastAppliedOffset() throws Exception {
                     singletonList(new ApiMessageAndVersion(new FeatureLevelRecord().
                         setName(MetadataVersion.FEATURE_NAME).
-                        setFeatureLevel(IBP_3_3_IV1.featureLevel()), (short) 0)),
+                        setFeatureLevel(IBP_3_3_IV3.featureLevel()), (short) 0)),
```
Since this can be any MV, could we just use MetadataVersion.MINIMUM_VERSION?

junrao commented on metadata/src/test/java/org/apache/kafka/image/loader/MetadataLoaderTest.java:
```diff
@@ -597,7 +597,7 @@ private void loadTestSnapshot2(
                         singletonList(new ApiMessageAndVersion(new FeatureLevelRecord().
                                 setName(MetadataVersion.FEATURE_NAME).
-                                setFeatureLevel(IBP_3_3_IV2.featureLevel()), (short) 0)),
+                                setFeatureLevel(IBP_3_4_IV0.featureLevel()), (short) 0)),
```
Since this can be any MV larger than MetadataVersion.MINIMUM_VERSION, could we just use MetadataVersion.latestProduction?

junrao commented on metadata/src/test/java/org/apache/kafka/image/MetadataVersionChangeTest.java:
```diff
@@ -30,31 +30,31 @@
 
-    private static final MetadataVersionChange CHANGE_3_0_IV1_TO_3_3_IV0 =
-        new MetadataVersionChange(IBP_3_0_IV1, IBP_3_3_IV0);
+    private static final MetadataVersionChange CHANGE_3_3_IV3_TO_3_6_IV0 =
```
We just need two different MVs here. Could we use MetadataVersion.MINIMUM_VERSION and MetadataVersion.latestProduction?

junrao commented on server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:
```diff
@@ -29,38 +29,20 @@
-            int expectedLevel = i - MetadataVersion.IBP_3_0_IV1.ordinal() + 1;
+    public void testFeatureLevels() {
+        for (int i = MINIMUM_VERSION.ordinal(); i < MetadataVersion.VERSIONS.length; i++) {
+            int expectedLevel = i + 7;
```
Could we use MINIMUM_VERSION.featureLevel() instead of 7?

junrao reviewed (commented):
@ijuma : Thanks for the updated PR. Made a pass of all files. A few more comments.

junrao reviewed (commented):
Also, there are a few references to 3.0 in docs/ops.html that need to be changed.

`Feature: metadata.version       SupportedMinVersion: 3.0-IV1    SupportedMaxVersion: 3.9-IV0 FinalizedVersionLevel: 3.9-IV0  Epoch: 5`

ijuma committed:
Address review comments

m1a2st commented on metadata/src/main/java/org/apache/kafka/controller/ActivationRecordsGenerator.java:
```diff
@@ -153,7 +153,7 @@ static ControllerResult<Void> recordsForNonEmptyLog(
+     * metadata transactions are supported, use them. Otherwise, write the bootstrap records as an
      * atomic batch. The single atomic batch can be problematic if the bootstrap records are too large
      * (e.g., lots of SCRAM credentials). If ZK migrations are enabled, the activation records will
      * include a ZkMigrationState record regardless of whether the log was empty or not.
```
Please also cleanup this outdate ZK migrations comment

m1a2st commented on core/src/main/scala/kafka/server/AlterPartitionManager.scala:
```diff
@@ -224,53 +215,35 @@ class DefaultAlterPartitionManager(
   private def buildRequest(
     inflightAlterPartitionItems: Seq[AlterPartitionItem],
     brokerEpoch: Long
```
Please also update the return value document.

m1a2st commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -552,24 +535,10 @@ class AlterPartitionManagerTest {
 
 object AlterPartitionManagerTest {
-  def provideMetadataVersions(): JStream[MetadataVersion] = {
+  def provideLeaderRecoveryState(): JStream[Arguments] = {
```
we can use `@EnumSource(classOf[LeaderRecoveryState])`, thus this method can remove

m1a2st commented on core/src/test/scala/unit/kafka/server/ApiVersionsResponseIntegrationTest.scala:
```diff
@@ -38,28 +37,18 @@ class ApiVersionsResponseIntegrationTest extends BaseRequestTest {
-    } else {
-      assertEquals(0, response.data().supportedFeatures().size())
-    }
+    assertFeatureHasMinVersion("metadata.version", response.data().supportedFeatures(), 7)
```
ditto `MINIMUM_VERSION.featureLevel()` instead of 7 ?

m1a2st commented on core/src/test/scala/unit/kafka/server/ApiVersionsResponseIntegrationTest.scala:
```diff
@@ -38,28 +37,18 @@ class ApiVersionsResponseIntegrationTest extends BaseRequestTest {
-    } else {
-      assertEquals(0, response.data().supportedFeatures().size())
-    }
+    assertFeatureHasMinVersion("metadata.version", response.data().supportedFeatures(), 7)
```
ditto `MINIMUM_VERSION.featureLevel()` instead of 7 ?

m1a2st commented on metadata/src/test/java/org/apache/kafka/controller/FeatureControlManagerTest.java:
```diff
@@ -99,10 +100,10 @@ public void testUpdateFeatures() {
             build();
         snapshotRegistry.idempotentCreateSnapshot(-1);
-        assertEquals(new FinalizedControllerFeatures(Collections.singletonMap("metadata.version", (short) 4), -1),
+        assertEquals(new FinalizedControllerFeatures(Collections.singletonMap("metadata.version", (short) 7), -1),
```
ditto `MINIMUM_VERSION.featureLevel()` instead of 7 ?

m1a2st commented on metadata/src/test/java/org/apache/kafka/controller/FeatureControlManagerTest.java:
```diff
@@ -142,11 +143,11 @@ public void testReplay() {
         manager.replay(record);
         snapshotRegistry.idempotentCreateSnapshot(123);
-        assertEquals(new FinalizedControllerFeatures(versionMap("metadata.version", 4, "foo", 2), 123),
+        assertEquals(new FinalizedControllerFeatures(versionMap("metadata.version", 7, "foo", 2), 123),
```
ditto `MINIMUM_VERSION.featureLevel()` instead of 7 ?

m1a2st reviewed (commented):
Thanks for @ijuma update, left some comments

ijuma committed:
Don't set metadata version to the `MINIMUM_VERSION` and rely only on the log to try and debug CI-only test failures

ijuma committed:
Add fallback in case we read a feature level we don't understand (for a specific range)

ijuma commented on metadata/src/main/java/org/apache/kafka/image/FeaturesDelta.java:
```diff
@@ -58,7 +59,14 @@ public void finishSnapshot() {
+            // We automatically fallback to IBP_3_3_IV3 in that case. We use explicit versions instead of `MINIMUM_VERSION` because
+            // we want to force an explicit decision if we change `MetadataVersion.MINIMUM_VERSION` in the future.
+            if (record.featureLevel() >= MINIMUM_PERSISTED_FEATURE_LEVEL && record.featureLevel() <= MetadataVersion.IBP_3_3_IV3.featureLevel())
+                metadataVersionChange = MetadataVersion.IBP_3_3_IV3;
```
@junrao @cmccabe Thoughts on this fallback? It seems like quite a rare edge case (since it requires usage of a pre-release version and for the older metadata version not to be compacted away if I understand correctly), but it seems fair to never fail while reading the log and the fallback code is pretty simple.

junrao replied:
Since we only support upgrading from officially released 3.3 to 4.0, we could probably just fail MV 3_3_IV0 to 3_3_IV2 as other unsupported MVs.

ijuma replied:
Is it guaranteed that we won't have a record with the older metadata version in the log in that case?

junrao replied:
There is no guarantee, but the expected MV is IBP_3_3_IV3 and above. If this is not the case, it seems easier to just fail instead of changing it silently.

ijuma replied:
I discussed this with a few people and the bit that was missing in the discussion above is that we generate snapshots every hour by default and hence this issue is unlikely to ever happen and the workaround is reasonably simple. I updated the code to remove the fallback and include a helpful message if this issue occurs.

junrao commented on metadata/src/test/java/org/apache/kafka/image/MetadataVersionChangeTest.java:
```diff
@@ -30,31 +32,31 @@
 
-    private static final MetadataVersionChange CHANGE_3_3_IV0_TO_3_0_IV1 =
-        new MetadataVersionChange(IBP_3_3_IV0, IBP_3_0_IV1);
+    private static final MetadataVersionChange CHANGE_3_6_IV0_TO_3_3_IV3 =
```
Could we change this to use MetadataVersion.MINIMUM_VERSION and MetadataVersion.latestProduction() too?

ijuma committed:
Simplify code based on code review

ijuma committed:
Remove metadata version default

ijuma committed:
More IBP_3_3_IV3 -> MINIMUM_VERSION

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -1650,13 +1642,7 @@ private QuorumController(
-                }
-                return ControllerResult.of(records, false);
-            },
+            () -> ControllerResult.of(List.of(new ApiMessageAndVersion(new NoOpRecord(), (short) 0)), false),
```
The javadoc above on `if the MetadataVersion supports it.` needs to be changed too.

junrao reviewed (commented):
@ijuma : Thanks for the updated PR. A few more comments.

m1a2st commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -495,8 +495,13 @@ BrokerFeature processRegistrationFeature(
+        // MetadataVersion has no default while the other features default to `0`
+        short finalized;
+        if (feature.name().equals(MetadataVersion.FEATURE_NAME))
+            finalized = finalizedFeatures.get(feature.name()).orElseThrow(() -> new RuntimeException());
```
I think we should throw a more specific exception or include a clear error message.

Update: This is already resolved :)

ijuma committed:
Reject broker registration without metadata.version and fix tests

m1a2st reviewed (commented):
Thanks for @ijuma update, one more comment

ijuma committed:
Update MetadataBatchLoader.resetToImage to set `hasSeenRecord` based on whether image is empty

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kafka-18601-assume-baseline-3.3-for-server

* apache-github/trunk: (36 commits)
  MINOR: Add KIP-848's metric to the doc (#18890)
  KAFKA-18772 Define share group config defaults for Docker (#18899)
  KAFKA-18733: Updating share group metrics (1/N) (#18826)
  KAFKA-18634: Fix ELR metadata version issues (#18680)
  KAFKA-18298 Fix flaky testConsumerGroupsDeprecatedConsumerGroupState and testConsumerGroups in PlaintextAdminIntegrationTest (#18513)
  MINOR: Marking testVerifyFetchAndCloseImplicit flaky (#18893)
  MINOR: Adjust javadoc to reflect the correct status of standby task TopicPartition (#18892)
  KAFKA-17182: Consumer fetch sessions are evicted too quickly with AsyncKafkaConsumer (#18795)
  KAFKA-16720: Support multiple groups in DescribeShareGroupOffsets RPC (#18834)
  KAFKA-18654[2/2]: Transction V2 retry add partitions on the server side when handling produce request. (#18810)
  MINOR: fix warn log message in Kafka Streams (#18878)
  KAFKA-18776: Fix flaky coordinator disconnect test & fix log level (#18866)
  KAFKA-17298: Update upgrade notes for 4.0 KIP-848 (#18756)
  MINOR bump setup-gradle (#18879)
  MINOR: Run javadoc as part of check task (#18863)
  MINOR: Updated share partition manager tests to close and other fixes (#18862)
  MINOR: Fix typo in ClusterControlManager (#18886)
  KAFKA-18728 Move ListOffsetsPartitionStatus to server module (#18807)
  MINOR: cleanup KStream JavaDocs (14/N) - stream-globalTable-left-join (#18867)
  MINOR: cleanup KStream JavaDocs (13/N) - stream-stream-outer-join (#18865)
  ...

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kafka-18601-assume-baseline-3.3-for-server

* apache-github/trunk:
  MINOR: Add release notes for Transactions Server Side Defense (KIP-890) (#18896)
  MINOR: TransactionManager logs the epoch bump less frequently. (#18895)
  MINOR: Mark IBP_4_0_IV3 as production ready! (#18902)

ijuma commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataBatchLoader.java:
```diff
@@ -100,7 +100,7 @@ public boolean hasSeenRecord() {
     public final void resetToImage(MetadataImage image) {
         this.image = image;
-        this.hasSeenRecord = true;
+        this.hasSeenRecord = !image.isEmpty();
```
This was an existing issue that somehow became easier to trigger due to these changes. Thanks to @mumrah for figuring out that this was the reason for the remaining failures.

ijuma committed:
Fix test failures

ijuma committed:
Make FeaturesImage.metadataVersion optional and fail-fast instead of silently doing the wrong thing

ijuma commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -437,18 +438,6 @@ class KRaftClusterTest {
   }
 
-  @Test
-  def testCreateClusterInvalidMetadataVersion(): Unit = {
```
It's no longer possible to set the bootstrap metadata version to lower than 3.3-IV3.

ijuma committed:
Minor tweak

ijuma committed:
More 3_3_IV3 -> MINIMUM_VERSION

ijuma committed:
Fix several tests

ijuma committed:
Fix ReconfigurableQuorumIntegrationTest and QuorumControllerTest by allowing FeaturesDelta to create a FeaturesImage with no metadata version

ijuma committed:
Minor clean-ups in `ReconfigurableQuorumIntegrationTest`

ijuma committed:
"3-3-IV3" -> `MINIMUM_VERSION.toString`

ijuma commented:
@junrao @chia7712 The tests for JDK 23 were all passing for the last run (the JDK 17 build failed, but it didn't look related to this PR).

I think there's only one outstanding issue: how to handle the case if we read an old metadata version from the log. I'll update the relevant comment later today. Does the rest look good to you?

ijuma committed:
Remove metadata version fallback from `FeaturesDelta` - snapshots happen frequently enough that it's not worth handling this edge case.

ijuma commented:
I resolved the outstanding item, so no known issues remain in this PR.

chia7712 commented on core/src/main/scala/kafka/server/AlterPartitionManager.scala:
```diff
@@ -92,8 +87,7 @@ object AlterPartitionManager {
       time = time,
       brokerId = config.brokerId,
-      brokerEpochSupplier = brokerEpochSupplier,
-      metadataVersionSupplier = () => metadataCache.metadataVersion()
```
we can remove `metadataCache` from the `apply` method

chia7712 commented on metadata/src/main/java/org/apache/kafka/image/writer/ImageWriterOptions.java:
```diff
@@ -29,30 +29,22 @@
             this.isEligibleLeaderReplicasEnabled = image.features().isElrEnabled();
         }
 
         public Builder setMetadataVersion(MetadataVersion metadataVersion) {
```
After removing those if-else, it seems we don't need the builder anymore. Maybe we can rewrite `ImageWriterOptions` by record class

ijuma replied:
This is a general pattern for the metadata module, I would rather not change it as part of this PR. We can have that discussion via a separate JIRA if useful.

chia7712 commented on metadata/src/test/java/org/apache/kafka/image/MetadataVersionChangeTest.java:
```diff
@@ -30,31 +30,31 @@
 
-    private static final MetadataVersionChange CHANGE_3_0_IV1_TO_3_3_IV0 =
-        new MetadataVersionChange(IBP_3_0_IV1, IBP_3_3_IV0);
+    private static final MetadataVersionChange CHANGE_MINUMUM_TO_LATEST =
```
typo: `MINUMUM` ->  `MINIMUM`

chia7712 commented on core/src/main/scala/kafka/server/AlterPartitionManager.scala:
```diff
@@ -114,7 +107,7 @@ class DefaultAlterPartitionManager(
   // use an AlterPartition request version which supports topic ids in the end because 
   // the two updates with the same name would be merged together.
-  private[server] val unsentIsrUpdates: util.Map[TopicPartition, AlterPartitionItem] = new ConcurrentHashMap[TopicPartition, AlterPartitionItem]()
+  private[server] val unsentIsrUpdates = new ConcurrentHashMap[TopicIdPartition, AlterPartitionItem]()
```
Could you please update the docs according to the change?

chia7712 commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -718,30 +710,14 @@ public void testRegistrationWithUnsupportedMetadataVersion() {
-    @Test
-    public void testRegisterControlWithOlderMetadataVersion() {
-        FeatureControlManager featureControl = new FeatureControlManager.Builder().
-            setMetadataVersion(MetadataVersion.IBP_3_3_IV0).
```
This test appears necessary as it verifies MV which is older than `IBP_3_7_IV0`.  Perhaps we could simply change `IBP_3_3_IV0` to `IBP_3_3_IV3` instead.

ijuma replied:
I deleted that because we have `testRegistrationWithUnsupportedMetadataVersion`. Does that address your concern?

chia7712 replied:
It appears that `testRegistrationWithUnsupportedMetadataVersion` is designed to test broker registration, while `testRegisterControlWithOlderMetadataVersion` is for controller registration.

ijuma replied:
Good catch, I had missed that. I'll add the test back with `IBP_3_6_IV2` since that's the newest IBP that still fails.

chia7712 commented on metadata/src/test/java/org/apache/kafka/image/writer/ImageWriterOptionsTest.java:
```diff
@@ -48,40 +48,33 @@
     }
 
     @Test
     public void testSetMetadataVersion() {
```
If we rewrite `ImageWriterOptions` by record class, this `testSetMetadataVersion` is unnecessary as all are pure setters.

ijuma replied:
I agree that this test isn't useful since the special logic for the setter was removed - even without removing the builder. I removed it.

chia7712 reviewed (commented):
@ijuma thanks for this patch. a couple of comments remain.

ijuma committed:
Address review feedback.

ijuma commented:
@chia7712 I addressed your comments.

chia7712 reviewed (approved):
LGTM

ijuma commented:
I will go ahead and merge this. If @junrao has any additional comments, I will submit a separate PR to address those.

mumrah commented:
Thanks for pushing this over the line @ijuma. It's great to see these cleanups and removals. "IBP" is all but a memory now :)
